kotlinx-coroutines-core/jvm/src/flow/internal/SafeCollector.kt

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2016-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow.internal
+
+import kotlinx.coroutines.flow.*
+import kotlin.coroutines.*
+import kotlin.coroutines.intrinsics.*
+import kotlin.coroutines.jvm.internal.*
+
+@Suppress("UNCHECKED_CAST")
+internal actual class SafeCollector<T> actual constructor(
+    @JvmField internal actual val collector: FlowCollector<T>,
+    @JvmField internal actual val collectContext: CoroutineContext
+) : FlowCollector<T> {
+
+    @JvmField // Note, it is non-capturing lambda, so no extra allocation during init of SafeCollector
+    internal actual val collectContextSize = collectContext.fold(0) { count, _ -> count + 1 }
+    private var lastEmissionContext: CoroutineContext? = null
+    private var intercepted: InterceptedContinuationOwner = InterceptedContinuationOwner()
+    private val emitFun = run {
```

Comment:
There's no need to use `run { ... }` and this whole `emitFun` should be a top-level constant in this file. Hint: you don't have to bind `collector` ahead of time, you can do:
```
private val emitFun =
    FlowCollector<*>::emit as Function3<FlowCollector<*>, Any?, Continuation<Unit>, Any?>
```
and the substitute an actual collector at the call-site.

Reply:
Good point. Though for wildcards compiler explicitly puts checkcast to `Void` class, so `Any?` should be used instead